lib/platform-stats.js

```diff
@@ -99,30 +93,155 @@ export const updateDailyStationStats = async (
 
 /**
  * @param {import('pg').Client} pgClient
- * @param {Set<string>} participants
+ * @param {number[]} participantIds
  */
-export const updateDailyParticipants = async (pgClient, participants) => {
-  debug('Updating daily participants, count=%s', participants.size)
-  const ids = await mapParticipantsToIds(pgClient, participants)
+export const updateDailyParticipants = async (pgClient, participantIds) => {
+  debug('Updating daily participants, count=%s', participantIds.length)
   await pgClient.query(`
     INSERT INTO daily_participants (day, participant_id)
     SELECT now() as day, UNNEST($1::INT[]) AS participant_id
     ON CONFLICT DO NOTHING
   `, [
-    ids
+    participantIds
+  ])
+}
+
+/**
+ * @param {import('pg').Client} pgClient
+ * @param {import('./preprocess.js').Measurement[]} honestMeasurements
+ * @param {import('./preprocess.js').Measurement[]} allMeasurements
+ * @param {Map<string, number>} participantsMap
+ * @param {object} options
+ * @param {Date} [options.day]
+ */
+export const updateStationsAndParticipants = async (
+  pgClient,
+  honestMeasurements,
+  allMeasurements,
+  participantsMap,
+  { day = new Date() } = {}
+) => {
+  /** @type {Map<number, Map<string, {accepted: number, total: number}>>} */
+  const participantsStationsStats = new Map()
+
+  /** @type {Map<number, Set<string>>} */
+  const subnets = new Map()
+
+  /** @type {(m: import('./preprocess.js').Measurement, participantId: number, countType: 'accepted' | 'total') => void} */
+  const updateParticipantsStationsStats = (m, participantId, countType) => {
+    let participant = participantsStationsStats.get(participantId)
+    if (!participant) {
+      participant = new Map()
+      participantsStationsStats.set(participantId, participant)
+    }
+
+    let stationStats = participant.get(m.stationId)
+    if (!stationStats) {
+      stationStats = { accepted: 0, total: 0 }
+      participant.set(m.stationId, stationStats)
+    }
+
+    stationStats[countType]++
+  }
+
+  for (const m of honestMeasurements) {
+    if (m.stationId == null) continue
+    const participantId = participantsMap.get(m.participantAddress)
+    if (participantId == null) continue
+
+    updateParticipantsStationsStats(m, participantId, 'accepted')
+  }
+
+  for (const m of allMeasurements) {
+    if (m.stationId == null) continue
+    const participantId = participantsMap.get(m.participantAddress)
+    if (participantId == null) continue
+
+    updateParticipantsStationsStats(m, participantId, 'total')
+
+    let subnetsSet = subnets.get(participantId)
+    if (!subnetsSet) {
+      subnetsSet = new Set()
+      subnets.set(participantId, subnetsSet)
+    }
+
+    subnetsSet.add(m.inet_group)
+  }
```

Comment:
Sounds good! The only downside I see is that we are spreading apart the logic that decides which measurements are accepted vs not, rather than having it in one single place. But it would make this simpler code, and we already do it elsewhere, so I'll make the change